Extensions/Leaderboards/leaderboardstools.ts

```diff
@@ -45,74 +116,196 @@ namespace gdjs {
+          }
+        ).then(
+          (response) => {
+            if (!response.ok) {
+              const errorCode = response.status.toString();
+              logger.error(
+                'Server responded with an error:',
+                errorCode,
+                response.statusText
+              );
+              scoreSavingData.setError(errorCode);
+              errorVar.setString(errorCode);
+              return;
             }
-          )
-            .then((response) => {
-              _scoreLastSentAt = Date.now();
-              if (!response.ok) {
-                errorVar.setString(response.status.toString());
-                _lastErrorCode = response.status;
-                return response.statusText;
-              } else {
-                _lastScore = score;
-                _lastPlayerName = playerName;
-                _lastErrorCode = response.status;
-                return response.text();
+
+            scoreSavingData.closeSaving();
+
+            response.text().then(
+              (text) => {
+                responseVar.setString(text);
+              },
+              (error) => {
+                logger.warn(
+                  'An error occurred when reading response but score has been saved:',
+                  error
+                );
+                responseVar.setString('CANNOT_READ_RESPONSE');
               }
-            })
-            .then((data) => {
-              responseVar.setString(data);
-            })
-            .catch((error) => {
-              logger.warn(
-                `Error while submitting a leaderboard score: ${error}`
-              );
-              errorVar.setString('REQUEST_NOT_SENT');
-              _lastErrorCode = 400;
-            });
+            );
+          },
+          (error) => {
+            logger.error('Error while submitting a leaderboard score:', error);
+            const errorCode = 'REQUEST_NOT_SENT';
+            scoreSavingData.setError(errorCode);
+            errorVar.setString(errorCode);
+          }
+        );
+      };
+
+      const getLeaderboardOrLastLeaderboardState = function (
+        state: string,
+        leaderboardId?: string
+      ): boolean {
+        if (leaderboardId) {
+          if (!_savingDataByLeaderboard[leaderboardId]) {
+            return false;
+          }
+          return _savingDataByLeaderboard[leaderboardId][state];
         }
+        const lastScoreSavingData = getLastScoreSavingData({
+          hasEnded: state === 'hasScoreBeenSaved',
+        });
+        if (!lastScoreSavingData) {
+          return false;
+        }
+        return lastScoreSavingData[state];
       };
 
-      export const hasLastEntrySaveFailed = function () {
-        return _lastErrorCode && _lastErrorCode >= 400;
+      export const isSaving = function (leaderboardId?: string): boolean {
+        return getLeaderboardOrLastLeaderboardState(
+          'isScoreSaving',
+          leaderboardId
+        );
+      };
```

Comment:
Prefer a typesafe equivalent:
```suggestion
      
      export const isSaving = function (leaderboardId?: string): boolean {
        const savingState = getLeaderboardOrLastLeaderboardSavingState(leaderboardId);
        return savingState ? savingState.isScoreSaving : false
      };
```